### PR TITLE
WIP: Add user defined aggregations

### PR DESCRIPTION
**Pull request by chmp**
As discussed in #2340, a first draft of user defined aggregations. The changes are minimal and hook nicely into the existing aggregation machinery. As a test, I reimplemented mean / sum as custom aggregations. 

A small caveat: the implementation is somewhat hacky, since it relys on `GroupBy.dtypes.columns` to get the columns of a grouped dataframe. However, I didn't see any other way, except to change the signature of all existing aggregations.

Missing are more tests in general and tests for dd.Series objects in particular. 

Also: it could make sense to reimplement the existing aggregations in terms of `Aggregation` to cleanup the code.

### CODE REVIEW

**Comment by mrocklin** [comment]
Hrm, interesting.  My first thought for this was a method that takes a per-chunk, aggregate, and finalize functions as arguments:

    df.groupby(...).custom(chunk, aggregate, finalize)

However I see now that this wouldn't easily generalize to the full `.aggregate` method except perhaps if we created a convention of passing a dict like `{'chunk': chunk, 'aggregate': aggregate}`.

Generally creating custom classes within Dask is rare, though this has probably been mostly due to an idiosyncratic aversion to classes of the people writing it so far, and not for any particularly strong reason.  The proposed implementation here seems reasonable.  It would be nice to get other peoples' opinions.

cc @TomAugspurger @sinhrks @jcrist

**Comment by chmp** [comment]
> Generally creating custom classes within Dask is rare, though this has probably been mostly due to an idiosyncratic aversion to classes of the people writing it so far, and not for any particularly strong reason. The proposed implementation here seems reasonable. It would be nice to get other peoples' opinions.

The reason I used a class here is a potential confusion when calling `.aggregate`: with dicts, you could not use `.aggregate(custom)`, but always have to use `.aggregate({col1: custom, col2: custom})`. I guess though it won't be a huge issue in everyday use.

File: dask/dataframe/groupby.py [hunk 1]
```diff
@@ -553,6 +624,19 @@ def _apply_func_to_column(df_like, column, func):
     return func(df_like[column])
 
 
+def _apply_func_to_columns(df_like, prefix, func):
+    if isinstance(df_like, pd.DataFrame):
+        columns = df_like.columns
+    else:
+        # TODO: this looks way to hacky
```
Comment (TomAugspurger) [inline dask/dataframe/groupby.py]: I think using `df_like._selected_obj.columns` should work for you.

File: dask/dataframe/groupby.py [hunk 2]
```diff
@@ -427,6 +464,9 @@ def _build_agg_args_single(result_column, func, input_column):
     elif func == 'mean':
         return _build_agg_args_mean(result_column, func, input_column)
 
+    elif isinstance(func, Aggregation):
```
Comment (chmp) [inline dask/dataframe/groupby.py]: Generally speaking, the aggregations can be arbitrarily named. Problems only arise, if a function name-clashes for the same column, since then intermediate values will be passed incorrectly and both will use the same output column . 

I added tests to clarify the current behavior and added a descriptive exception for the error case when `.aggregate` is called.

**Comment by mrocklin** [comment]
What is the status here.  @TomAugspurger are you ok with this PR?

**Comment by TomAugspurger** [comment]
Looks like @chmp addressed all my prior comments. @chmp one more ask: could you add an example in `doc/source/dataframe-groupby.rst`? That would be useful to link to when questions come up.

**Comment by chmp** [comment]
Good idea. I'll add an example in the next days.

File: dask/dataframe/groupby.py [hunk 3]
```diff
@@ -1044,17 +1144,13 @@ def nunique(self, split_every=None, split_out=1):
 
     @derived_from(pd.core.groupby.SeriesGroupBy)
     def aggregate(self, arg, split_every=None, split_out=1):
```
Comment (chmp) [inline dask/dataframe/groupby.py]: The short circuit-behavior is no longer necessary, since this is addressed by the added code. For `sum, size, count` the old implementation should be faster. For `var`, `mean`, `std`, the new implementation should be faster (less conversions). In all cases, this removes one additional branch to keep track of. I'm happy to re-add the original code, if you'd prefer.

**Comment by chmp** [comment]
I added examples for custom aggregations and a general introduction to `.aggregate`. 

I also tested, and then fixed, the behavior for series `.aggregate` with a single argument. In changing this code, I also removed an extra if-branch, that is no longer necessary (see comment above).

**Comment by mrocklin** [comment]
This looks like a nice change to me.  It's currently failing a test and has a TODO for examples.  I would like to release sometime in the near future.  If we could get this in before then that would be ideal (though waiting is also ok if people are busy).

**Comment by chmp** [comment]
I have some time this week and will try to finish everything up. 

The failing tests are interestingly related to the pandas version, with pandas `20.0.1` everything works. I'll look into it and fix it. I will also add the promised examples ;). Should I also add `dd.Aggregation` to the API docs somewhere? If so, some pointers for an appropriate place would be appreciated.

**Comment by chmp** [comment]
I opted to change the test to ignore the name property. The failure is due to an inconsistency in the pandas API that was fixed with pandas `0.20.0`. The following script demonstrate the problem:

```python
import pandas as pd
df = pd.DataFrame({'g': [0, 1], 'v': [1, 2]})

print(
    'version:', pd.__version__,
    'mean: ', df.groupby('g')['v'].mean().name,
    'size:', df.groupby('g')['v'].size().name,
)
```

Depending on the pandas version I get `version: 0.19.0 mean: v size: None` or `version: 0.20.2 mean: v size: v`.

I also added the documentation to the docstring of `dd.Aggregation`.

**Comment by mrocklin** [comment]
Now that `pandas.DataFrame.agg` exists should we try to make this aggregation work in either case?

**Comment by chmp** [comment]
`dd.Dataframe.agg` would have been my next target :). However, I would prefer to open a new ticket / PR. If you're worried about compatibility, I don't see any reason the proposed interface for custom aggregations should not work for `df.agg(...)`, since it is in essence an optimization of `df['__one__'] = 1; df.groupby('__one__').agg(...)`.

**Comment by chmp** [comment]
On the other hand, maybe the changes are not too involved. I'll give a it a spin and see how much would need to change.

**Comment by TomAugspurger** [comment]
@chmp do you have a chance to revisit this? I think holding off on `dataframe.agg` for now is just fine. IIRC, the groupby version was essentially finished.

**Comment by chmp** [comment]
I rebased based to master and everything seems to work. However, one [travis job](https://travis-ci.org/dask/dask/jobs/271423599) failed with multiple test errors. As far as I can tell, it is caused by a crashed worker and not related to my code. Could you have a look by any chance?

Apart form that the state is the same as it was before: 

- `df.groupby(...).agg` and `series.groupy(..).agg` support custom aggregations
- `df.agg` does not. I looked into it, but ran into a couple of issues and lost unfortunately my work-in progress code.

As far as I can tell, there are no open issues, assuming my changes did not cause the test failures, are there?

**Comment by TomAugspurger** [comment]
Thanks. I restarted that job.

File: dask/dataframe/groupby.py [hunk 4]
```diff
@@ -163,6 +164,54 @@ def _groupby_get_group(df, by_key, get_key, columns):
...
+        a function that will be called to aggregate the results of each chunk.
+        Again the argument(s) will be grouped series. If ``chunk`` returned a
+        tuple, ``agg`` will be called with all of them as individual positional
+        arguments.
+
+    finalize: callable
+        an optional finalizer that will be called with the results from the
+        aggregation.
+
+    For example, ``sum`` can be implemented as::
```
Comment (TomAugspurger) [inline dask/dataframe/groupby.py]: Put this under an `Examples` header.

File: dask/dataframe/groupby.py [hunk 5]
```diff
@@ -163,6 +164,54 @@ def _groupby_get_group(df, by_key, get_key, columns):
...
+        custom_sum = dd.Aggregation('custom_sum', lambda s: s.sum(), lambda s0: s0.sum())
+        df.groupby('g').agg(custom_sum)
+
+    and ``mean`` can be implemented as::
+
+        custom_mean = dd.Aggregation(
+            'custom_mean',
+            lambda s: (s.count(), s.sum()),
+            lambda count, sum: (count.sum(), sum.sum()),
+            lambda count, sum: sum/ count,
```
Comment (TomAugspurger) [inline dask/dataframe/groupby.py]: Space before the `/`: `sum / count`

File: docs/source/dataframe-groupby.rst [hunk 6]
```diff
@@ -104,3 +104,49 @@ keyword argument to ``groupby`` or ``set_index``
...
+these reductions, the steps should return tuples and expect multiple arguments.
+A mean function can be implemented as
+
+.. code-block:: python
+
+    custom_mean = dd.Aggregation(
+        'custom_mean',
+        lambda s: (s.count(), s.sum()),
+        lambda count, sum: (count.sum(), sum.sum()),
+        lambda count, sum: sum/ count,
```
Comment (TomAugspurger) [inline docs/source/dataframe-groupby.rst]: Space before the `/`

**Comment by chmp** [comment]
I addressed your comments and fixed the formatting issues.

**Comment by TomAugspurger** [comment]
Thanks @chmp!

**Comment by samuelroberth** [comment]
Thank @chmp! This is very valuable to me.

**Comment by chmp** [comment]
@TomAugspurger Thanks for your help in polishing the code and getting it merged.

@samuelroberth Cool. Happy to help. :)

**Comment by mrocklin** [comment]
@chmp you may be interested in this Stack Overflow question: https://stackoverflow.com/questions/46080171/constructing-mode-and-corresponding-count-functions-using-custom-aggregation-fun

**Comment by chmp** [comment]
@mrocklin, thanks. I am not that active on stack overflow and would not have noticed this question, had you not brought it my attention. I posted an [implementation](https://stackoverflow.com/a/46082075/8570642) of the mode function. 

Probably the current docs of the  are not the easiest to follow. It would probably make sense to adapt the docs depending on the feedback from SO. Maybe show the simplified logic behind the aggregate API:

```python
res = chunk(df.groupby('g')['col'])
res = agg(res.groupby(level=[0]))
res = finalize(res)
```
Potentially add the mode function to the examples. Although it is a bit wordy with explanation.
